### ISSUE TITLE
support for new t2.micro

### ISSUE THREAD

**Issue by jirubio** [aws/opsworks-cookbooks#162]
Hi, I migrated one t1.micro to the new t2.micro and found some problems.
Installing the built-in rails layer fails when installing passenger saying there is no more memory. After some research I find out that in the case of the t1.micro there is a recipe that creates a swap file to increase the memory. This recipe has the instance type hardcoded to "t1.micro". I tried to add the swap file manually to the t2.micro and everything works now.

There are some recipes that have the micro instance hardcoded, theses places should be modified to include the new t2.micro. Other option would be to make the swap instance type configurable via an attribute.

**Comment by jbraeuer** [comment]
Thanks for reporting this issue. I can confirm, that swap file is only created on t1.micro instances.

**Comment by gretel** [comment]
i'd love to choose wether the instance has swap space or not.

**Comment by jweiss** [comment]
@gretel Would you really want to disable swap on a T1 or T2? They are running into problems during the Chef run otherwise...

Or do you mean in general?

**Comment by gretel** [comment]
@jweiss, yes, in general.

**Comment by jirubio** [comment]
One easy solution is to create a new attribute that holds a set of instance types like "t1.micro;t2.micro" for which we want to create the swap file. And then use this attribute in the if to decide whether to apply the swap recipe or not. This would be the default behaviour, but if someone wants to add swap file to other instances or amazon releases new instance types you only have to override the attribute in the customize.rb file.

I searched the opsworks repository for the string "t1.micro" and there are a couple of places where it appears, so beware it's not only the swap thing (although this one is a blocker).

**Comment by jbraeuer** [comment]
Thanks for the input @jirubio. Having a attribute is a good suggestion, but it should be an Array instead of a String.
Do you have certain use-cases that require you to have swap enabled on instances other than t1/t2.micro?

**Comment by jirubio** [comment]
I have read somewhere that there were some problems when merging array attributes, this is why I proposed the string variant. Anyway I'm not a chef expert.
At this moment I don't have any use case for having to use swap in other instances than the micros. I think there should be problems when using instances with 1GB or less of ram, at least with passenger. So most probably this will only affect the micros.

**Comment by jbraeuer** [comment]
Support for swap on t2.micro instances was added with 7b861a2beba94624c2d7aeb9009c2cd1257f688e (Chef 11.10) and d858fce2849ff48169987adbead2532908d85570 (Chef 11.4).

Closing the issue.

**Comment by khobbits** [comment]
Just adding a note here in case it helps anyone else (I stumbled in here from Google).

I was attempting to host an application which requires a large amount of swap space on a t2.medium instance and had a few issues.  I solved this by following a few steps:
- Creating a ebs volume and mounting it at /var/swapdisk
- Adding the "opsworks_initial_setup::swap" recipe to the Custom Chef setup steps
- Adding the following stack json:

```
  "opsworks_initial_setup": {
    "swapfile_name": "/var/swapdisk/swapfile",
    "swapfile_size_mb": 4096
  },
```

If you try to set the instance type, the initial setup recipe runs before the ebs volume is mounted, and setup fails.
